[Prototype] Rework events (support DI)

HaoK opened the pull request:
- Nukes all IAuthEvents interfaces
- Introduces AuthenticationEvents base which contains Type EventsType which causes DI replacement of the events instance if set

Seems fairly seemless, no tests needed to be updated with this change surprisingly...  I added one for cookies to pull its events from DI

@Tratcher @PinpointTownes @davidfowl

dnfclas commented:
Hi __@HaoK__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

PinpointTownes commented:
Getting rid of the interfaces would be a cool change as we could introduce new events without breaking the contract... but are you sure it's really non-breaking? (I'm tempted to think it's a binary change :fearful:)

HaoK commented:
@PinpointTownes this is meant to be a breaking change :)

PinpointTownes commented:
@HaoK ah, it would have to wait until 2.0 then, right?

HaoK commented:
Yes but we are just validating the design with a prototype right?

HaoK commented:
Removed the one time resolution of the events type from the base middleware, and updated JwtBearer and cookie to demonstrate the new pattern:

Each events class will have (yet another) new method that hides a `ResolveEvents(IServiceProvider services)` that will replace or return itself.  The handlers will all use this method with context.RequestServices instead of directly using Options.Events

PinpointTownes commented:
> Yes but we are just validating the design with a prototype right?

Sure. But removing the interfaces is orthogonal. The "events type" thingy could be added in a non-breaking way, no?

> Removed the one time resolution of the events type from the base middleware, and updated JwtBearer and cookie to demonstrate the new pattern:

So, to make things clearer (for me :smile:), to use a specific events class, I'd have to do something like that:

```
app.UseJwtBearerAuthentication(new JwtBearerOptions {
    Events = new JwtBearerEvents {
        EventsType = typeof(MySuperEventsClass)
    }
});
```

... right?

It seems odd to me that you have to instantiate a `JwtBearerEvents` object to tell the system "hey, yeah, I provide you with this events class instance, but actually, you won't really use it. Instead, you'll resolve the DI-injected service represented by `EventsType`".

HaoK commented:
Sure but this PR is taking a look at all the events changes we'd like to make, not trying to scope this down since this is more of a prototype for now...

In regards to the usage, its a bit clunky due to no lambda so you can't inline:
```
var options = new JwtBearerOptions();
options.Events.EventsType = typeof(MySuperEventsClass);
app.UseJwtBearerAuthentication(options));
```

PinpointTownes commented:
In this case, we could also reconsider https://github.com/aspnet/Options/issues/11 for 2.0.

HaoK commented:
Not sure we want to add anything to options regarding this anymore, but we could certainly see if there's anything that makes sense to add to DI to make this pattern easier.

Tratcher commented on src/Microsoft.AspNetCore.Authentication/Events/AuthenticationEvents.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Base class for all authentication events.
+    /// </summary>
+    public class AuthenticationEvents
```
Using this as a base class isn't helpful when everybody just overrides it (and awkwardly at that).

HaoK commented:
Will be replaced as part of https://github.com/aspnet/Security/pull/1065
